File: antiarithmetic.cpp
#include <bits/stdc++.h>
using namespace std;

bool check(int p[], int n){
    bool flag = true;
    int t, h, k, l;
    for(int i=0; i<n-2; i++){
    	t = p[i];
        for(int j=i+1; j<n; j++){
        	h = p[j];
        	k = 2*j - i;
        	if(k >= n){
            	break;
            }
            l = p[k];
            if((t < h and h < l) || (l < h and h < t)){
            	// cout<<i<<" "<<j<<" "<<k<<"\n";
                flag = false;
                break;
            }
            
        }
        if(!flag){
            break;
        }
    }
    return flag;
}

int main(){
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);
    string s;
    while(1){
        cin>>s;
        if(s == "0"){
            break;
        }
        int n = stoi(s.substr(0, s.length()-1));
        int arr[n], pos[n], kk;
        for(int i=0; i<n; i++){
            cin>>kk;
            arr[i] = kk;
            pos[kk] = i;
        }
        if(check(pos, n)){
            cout<<"yes\n";
        }
        else{
            cout<<"no\n";
        }
    }
}

File: antiarithmetic.py
def check(l, n):
    p = [0 for i in range(n)]
    for i in range(n):
    	p[l[i]] = i 
    for i in range(n-2):
        for j in range(i+1, n):
        	k = 2*j - i
        	if k >= n:
        		break
        	if (p[i] > p[j] and p[j] > p[k]) or (p[k]>p[j] and p[j]>p[i]):
        		return False 
    return True 

while True:
    s = input()
    if s == '0':
        break 
    n, l = s.split(':')
    print(['no', 'yes'][check(list(map(int, l.split())), int(n))])
